### ISSUE TITLE
Reducing the size of the cuda_compiler zip

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-forge-pinning-feedstock#6967]
The size of the zip involving the cuda compilers is pretty unwieldy:
https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/ff043bb6eec88acf5c20bebccc255fc585014fc3/recipe/conda_build_config.yaml#L181-L189
which makes it hard to override locally. There's a general desire to keep the zip complexity down and remove entries here, rather than adding them.

And indeed we recently removed `cdt_name` (48621d1b6e8c44b77666cf3e36d3495085c86f1d) as well as `c_stdlib_version` (c7f59738d8f855541b0ec0b018251e18303f5310), and we made it easier to override the image version without having to locally override the whole zip (#6626).

#6910 now added another component (`github_action_labels`) to the zip, due to the fact that there exist feedstocks that require GPU agents (a very rare resource), and that without linking this with the zip that determines non-CUDA vs. CUDA, we're stuck with only bad options:
* waste resources by using GPU agents even for the non-CUDA builds
* local workarounds like https://github.com/h-vetinari/pytorch-cpu-feedstock/commit/6c232ae45e28059ef8c70ff36e6ea6d1cd601433, which cause several issues:
  * duplicated builds in CI matrix
  * very long variant config names, which causes failures on windows, see https://github.com/conda-forge/pytorch-cpu-feedstock/pull/332 & https://github.com/conda-forge/conda-smithy/pull/2233

The desire for simplity in the zip is well-founded, but it doesn't trump reality. Keeping these data apart when they are closely connected causes far more problems than a single extra zip entry (aside from the fact that only 5 feedstocks across all of conda-forge are affected and the required changes are trivial).

The CUDA zip overall is IMO a reflection of irreducible complexity that exists in our packaging constraints. However, there are some nice opportunities to further simplify this zip:
* `docker_image` can be removed as soon as we drop CUDA 11.8 (because then everything can just run on the alma9 images)
* `{c,cxx,fortran}_compiler_version` could be dropped if we accept being limited by the maximum GCC version supported by our lowest-supported CUDA
  * this can obviously be accelerated if we keep only building for the newest CUDA 12.x, as we recently started [doing](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6720) (and assuming that newer nvcc keeps updating their GCC compatibility).

If we did all that, the zip would reduce to
```yaml
zip_keys:
  -                             # [unix]
    - c_compiler_version        # [unix]
    - cxx_compiler_version      # [unix]
    - fortran_compiler_version  # [unix]
  # uncoupled!
  -                             # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
    - cuda_compiler             # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
    - cuda_compiler_version     # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
    - github_actions_labels     # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
```
which should be within the realm of acceptable zip size.

**Comment by isuruf** [comment]
> duplicated builds in CI matrix

This IMO is a bug.

> very long variant config names, which causes failures on windows, see
https://github.com/conda-forge/pytorch-cpu-feedstock/pull/332 & https://github.com/conda-forge/conda-smithy/pull/2233

This is also a bug in conda-smithy. We have
```
C:\Users\runnerx\_work\pytorch-cpu-feedstock\pytorch-cpu-feedstock\.ci_support\linux_64_blas_implgenericc_compiler_version13channel_targetsconda-forge_maincuda_compilerNonecuda_compiler_versionNonecxx_compiler_version13github_actions_labelscirun-openstack-cpu-2xlargeis_h2ab6b548.yaml
```
which has `284` characters and you can have only 260 on windows.

**Comment by isuruf** [comment]
> duplicated builds in CI matrix

I can help you with debugging this if you have a commit I can have a look at.

**Comment by h-vetinari** [comment]
> I can help you with debugging this if you have a commit I can have a look at.

It's just regular smithy behaviour. It could probably be avoided by adding
```yaml
  skip: true  # [cuda_compiler_version == "None" and github_actions_labels == "cirun-openstack-gpu-2xlarge"]
  skip: true  # [cuda_compiler_version != "None" and github_actions_labels == "cirun-openstack-cpu-2xlarge"]
```
to every single output, but that becomes another pointless thing to keep in sync.

> This is also a bug in conda-smithy. We have
> 
> ```
> C:\Users\runnerx\_work\pytorch-cpu-feedstock\pytorch-cpu-feedstock\.ci_support\linux_64_blas_implgenericc_compiler_version13channel_targetsconda-forge_maincuda_compilerNonecuda_compiler_versionNonecxx_compiler_version13github_actions_labelscirun-openstack-cpu-2xlargeis_h2ab6b548.yaml
> ```
> 
> which has `284` characters and you can have only 260 on windows.

ah, I was wondering how we got beyond 260 (https://github.com/conda-forge/conda-smithy/pull/2233 uses 200), as that was not at all obvious from the error message. However, if we subtract another 24 characters (to get that path <=260), we'll lose visibility of the GHA label in the path.

To be clear, I'm sure there's some halfway workable alternative to putting `github_action_labels` in the zip - but all those hacks are IMO worse than just dealing with the fact that we should be able to determine CPU vs. GPU agents together with the quantity (`cuda_compiler_version`) that determines whether or not we have a CUDA build.

All in all, the zip got shorter since last November, and it'll get shorter still once we drop CUDA 11.8. I don't see how a cosmetic concern outweighs all the workarounds that become necessary otherwise.

**Comment by isuruf** [comment]
> It's just regular smithy behaviour. It could probably be avoided by adding

Here's the recipe I tried and it skipped windows just like I intended

```yaml
package:
  name: foo-split
  version: 1.0.0

build:
  number: 1
  skip: true   # [win]

outputs:
  - name: foo

  - name: bar
    build:
      skip: true  # [linux]

about:
  home: asd
```

**Comment by h-vetinari** [comment]
See https://github.com/conda/conda-build/issues/4206

**Comment by h-vetinari** [comment]
> Here's the recipe I tried and it skipped windows just like I intended

You could also make a fork of https://github.com/conda-forge/tensorflow-feedstock/pull/414, and figure what's necessary to get smithy to only produce one set of jobs.

**Comment by h-vetinari** [comment]
@isuruf, did you have some time to look at how to achieve this (e.g. on the tensorflow feedstock)?

**Comment by h-vetinari** [comment]
@isuruf, the fact that we're wasting the GPU-enabled runners on CPU-only jobs is still an ongoing headache on the pytorch feedstock, especially if the server is congested. The zip in question has now been reduced to

https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/816165888306b37dbaf574adf21296252440629a/recipe/conda_build_config.yaml#L152-L157

I think one can hardly argue that adding one more item (`github_actions_labels`) to that zip would be "getting out of hand", given that this is how it looked like less than a year ago:

https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/f5b0a31b9774e485f5317279b9705d67b179779c/recipe/conda_build_config.yaml#L163-L172

Since you haven't followed up on your "I can help you with debugging this" for half a year, can we agree to move forward with a refresh of https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/ff043bb6eec88acf5c20bebccc255fc585014fc3 please?

(I'd still like to uncouple the GCC versions from `cuda_compiler_version`, but for that we'll have to drop CUDA on ppc64le)

**Comment by isuruf** [comment]
> for every output is not a workable solution anyway, because it's both cumbersome and fragile.

I meant that, it is not the problem. There's something else going wrong in conda-smithy/conda-build.

**Comment by h-vetinari** [comment]
I don't understand what you mean. Unless zipped, conda forms the cartesian product of keys with multiple values. That is consistent with what I'm seeing. Could you open a PR to pytorch or tensorflow to demonstrate what's going wrong according to you?

**Comment by h-vetinari** [comment]
We continue to experience more and more congestion on the opengpu server, and it's just a frivolous waste of resources to use GPU-enabled runners (the rarest resource we have) for building CPU-only packages. This is making an already tight bottleneck even tighter, and it's ridiculous to block improvements on this over a purely cosmetic issue (i.e. how large the zip involving `cuda_compiler_version` is).

@isuruf, you've blocked improvements on this for 10 months now, and not provided any of the promised alternative workarounds. You need to drop your objection here. Unless resolved by then, I will bring this up in the next core call.
